code/modules/surgery/operations/_operation.dm

```diff
@@ -0,0 +1,1280 @@
+			operation_args[OPERATION_SPEED] = 0
+
+		if(operation_args[OPERATION_FORCE_FAIL] || prob(clamp(GET_FAILURE_CHANCE(time, operation_args[OPERATION_SPEED]), 0, 99)))
+			failure(operating_on, surgeon, tool, operation_args)
+			result |= ITEM_INTERACT_FAILURE
+			update_surgery_mood(patient, SURGERY_STATE_FAILURE)
+		else
+			success(operating_on, surgeon, tool, operation_args)
+			result |= ITEM_INTERACT_SUCCESS
+			update_surgery_mood(patient, SURGERY_STATE_SUCCESS)
+
+		if(isstack(tool))
+			var/obj/item/stack/tool_stack = tool
+			tool_stack.use(1)
+
+	while ((operation_flags & OPERATION_LOOPING) && can_loop(patient, operating_on, surgeon, tool, operation_args))
+
+	SEND_SIGNAL(patient, COMSIG_LIVING_SURGERY_FINISHED, src, operating_on, tool)
+
+	if(patient.stat == DEAD && was_sleeping)
+		surgeon.client?.give_award(/datum/award/achievement/jobs/sandman, surgeon)
+
+	return result
+
+/// Called after an operation to check if it can be repeated/looped
+/datum/surgery_operation/proc/can_loop(mob/living/patient, atom/movable/operating_on, mob/living/surgeon, tool, list/operation_args)
+	PROTECTED_PROC(TRUE)
+	return operate_check(patient, operating_on, surgeon, tool, operation_args)
+
+/// Called during the do-after to check if the operation can continue
+/datum/surgery_operation/proc/operate_check(mob/living/patient, atom/movable/operating_on, mob/living/surgeon, tool, list/operation_args)
+	PROTECTED_PROC(TRUE)
+
+	if(isstack(tool))
+		var/obj/item/stack/tool_stack = tool
+		if(tool_stack.amount <= 0)
+			return FALSE
+
+	if(!surgeon.surgery_check(tool, patient))
+		return FALSE
+
+	if(!check_availability(patient, operating_on, surgeon, tool, operation_args[OPERATION_TARGET_ZONE]))
+		return FALSE
+
+	return TRUE
+
+/**
+ * Allows for any extra checks or setup when the operation starts
+ * If you want user input before for an operation, do it here
+ *
+ * This proc can sleep, sanity checks are automatically performed after it completes
+ *
+ * Return FALSE to cancel the operation
+ * Return TRUE to continue
+ */
+/datum/surgery_operation/proc/pre_preop(atom/movable/operating_on, mob/living/surgeon, tool, list/operation_args)
+	PROTECTED_PROC(TRUE)
+	return TRUE
+
+/// Used to display messages to the surgeon and patient
+/datum/surgery_operation/proc/display_results(mob/living/surgeon, mob/living/target, self_message, detailed_message, vague_message, target_detailed = FALSE)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	PROTECTED_PROC(TRUE)
+
+	surgeon.visible_message(detailed_message, self_message, vision_distance = 1, ignored_mobs = target_detailed ? null : target)
+	if(target_detailed)
+		return
+	var/you_feel = pick("a brief pain", "your body tense up", "an unnerving sensation")
+	if(!vague_message)
+		if(detailed_message)
+			stack_trace("DIDN'T GET PASSED A VAGUE MESSAGE.")
+			vague_message = detailed_message
+		else
+			stack_trace("NO MESSAGES TO SEND TO TARGET!")
+			vague_message = span_notice("You feel [you_feel] as you are operated on.")
```

Comment:
Traces probably should be more descriptive/specific if you want them to be helpful beyond just telling us that something broke, assuming tis is a debug/todo leftover?